File: vspace/src/main/java/edu/asu/diging/vspace/core/util/CitationFormatter.java
package edu.asu.diging.vspace.core.util;

import java.util.List;
import edu.asu.diging.vspace.core.model.impl.Reference;

public class APACitationFormatter {
    
    /**
     * Formats a list of references in APA style
     * 
     * @param references List of Reference objects to format
     * @return HTML formatted string with APA-style references
     */
    public static String formatReferences(List<Reference> references) {
        if (references == null || references.isEmpty()) {
            return "";
        }
        
        StringBuilder result = new StringBuilder();
        result.append("<div class=\"apa-references\">");
        
        for (int i = 0; i < references.size(); i++) {
            Reference ref = references.get(i);
            result.append(formatSingleReference(ref, i + 1));
        }
        
        result.append("</div>");
        return result.toString();
    }
    
    /**
     * Formats a single reference in APA style
     */
    private static String formatSingleReference(Reference ref, int index) {
        StringBuilder citation = new StringBuilder();
        citation.append("<div class=\"apa-reference\" id=\"ref-").append(index).append("\">");
        
        // Author(s)
        String authors = formatAuthors(ref.getAuthor());
        if (authors != null && !authors.isEmpty()) {
            citation.append(authors).append(" ");
        }
        
        // Year
        if (ref.getYear() != null && !ref.getYear().isEmpty()) {
            citation.append("(").append(ref.getYear()).append("). ");
        }
        
        // Title
        if (ref.getTitle() != null && !ref.getTitle().isEmpty()) {
            String title = ref.getTitle().trim();
            if (!title.endsWith(".")) {
                title += ".";
            }
            citation.append("<em>").append(title).append("</em> ");
        }
        
        // Journal/Publication info
        if (ref.getType() != null) {
            String type = ref.getType().toLowerCase();
            switch (type) {
                case "journal article":
                    formatJournalArticle(ref, citation);
                    break;
                case "book":
                    formatBook(ref, citation);
                    break;
                case "thesis":
                    formatThesis(ref, citation);
                    break;
                case "patent":
                    formatPatent(ref, citation);
                    break;
                case "report":
                    formatReport(ref, citation);
                    break;
                default:
                    formatGeneric(ref, citation);
                    break;
            }
        } else {
            formatGeneric(ref, citation);
        }
        
        // URL
        if (ref.getUrl() != null && !ref.getUrl().isEmpty()) {
            citation.append(" Retrieved from ").append(ref.getUrl());
        }
        
        citation.append("</div>");
        return citation.toString();
    }
    
    /**
     * Formats author names in APA style
     */
    private static String formatAuthors(String author) {
        if (author == null || author.isEmpty()) {
            return "";
        }
        
        String[] authors = author.split(",");
        StringBuilder formatted = new StringBuilder();
        
        for (int i = 0; i < authors.length; i++) {
            String name = authors[i].trim();
            if (name.isEmpty()) continue;
            
            String[] parts = name.split("\\s+");
            if (parts.length >= 2) {
                // Last name, First initial
                formatted.append(parts[parts.length - 1]).append(", ");
                for (int j = 0; j < parts.length - 1; j++) {
                    if (parts[j].length() > 0) {
                        formatted.append(parts[j].charAt(0)).append(". ");
                    }
                }
            } else {
                formatted.append(name);
            }
            
            if (i < authors.length - 1) {
                formatted.append(", ");
            }
        }
        
        return formatted.toString().trim();
    }
    
    /**
     * Formats journal article references
     */
    private static void formatJournalArticle(Reference ref, StringBuilder citation) {
        if (ref.getJournal() != null && !ref.getJournal().isEmpty()) {
            citation.append("<em>").append(ref.getJournal()).append("</em>");
        }
        
        if (ref.getVolume() != null && !ref.getVolume().isEmpty()) {
            citation.append(", ").append(ref.getVolume());
        }
        
        if (ref.getIssue() != null && !ref.getIssue().isEmpty()) {
            citation.append("(").append(ref.getIssue()).append(")");
        }
        
        if (ref.getPages() != null && !ref.getPages().isEmpty()) {
            citation.append(", ").append(ref.getPages());
        }
        
        citation.append(".");
    }
    
    /**
     * Formats book references
     */
    private static void formatBook(Reference ref, StringBuilder citation) {
        if (ref.getEditors() != null && !ref.getEditors().isEmpty()) {
            citation.append("(Ed.), ");
        }
        
        if (ref.getVolume() != null && !ref.getVolume().isEmpty()) {
            citation.append("Vol. ").append(ref.getVolume()).append(". ");
        }
        
        if (ref.getPages() != null && !ref.getPages().isEmpty()) {
            citation.append("pp. ").append(ref.getPages()).append(". ");
        }
    }
    
    /**
     * Formats thesis references
     */
    private static void formatThesis(Reference ref, StringBuilder citation) {
        citation.append("Unpublished ");
        if (ref.getJournal() != null && !ref.getJournal().isEmpty()) {
            citation.append(ref.getJournal()).append(" ");
        }
        citation.append("thesis.");
    }
    
    /**
     * Formats patent references
     */
    private static void formatPatent(Reference ref, StringBuilder citation) {
        citation.append("U.S. Patent No. ");
        if (ref.getVolume() != null && !ref.getVolume().isEmpty()) {
            citation.append(ref.getVolume());
        }
        citation.append(".");
    }
    
    /**
     * Formats report references
     */
    private static void formatReport(Reference ref, StringBuilder citation) {
        if (ref.getJournal() != null && !ref.getJournal().isEmpty()) {
            citation.append("(").append(ref.getJournal()).append("). ");
        }
        citation.append("Report.");
    }
    
    /**
     * Formats generic references
     */
    private static void formatGeneric(Reference ref, StringBuilder citation) {
        if (ref.getJournal() != null && !ref.getJournal().isEmpty()) {
            citation.append(ref.getJournal()).append(".");
        }
        
        if (ref.getVolume() != null && !ref.getVolume().isEmpty()) {
            citation.append(" ").append(ref.getVolume());
        }
        
        if (ref.getIssue() != null && !ref.getIssue().isEmpty()) {
            citation.append("(").append(ref.getIssue()).append(")");
        }
        
        if (ref.getPages() != null && !ref.getPages().isEmpty()) {
            citation.append(", ").append(ref.getPages());
        }
        
        citation.append(".");
    }
} 

File: vspace/src/main/webapp/resources/extra/citationFormatter.js
var VSpaceCitation = (function() {

    /**
     * Maps Virtual Spaces reference types to CSL-JSON types
     */
    var typeMapping = {
        'Book': 'book',
        'Journal Article': 'article-journal',
        'Conference Paper': 'paper-conference',
        'Thesis': 'thesis',
        'Report': 'report',
        'Website': 'webpage',
        'Patent': 'patent',
        'Other': 'document'
    };

    /**
     * Parses author string from "LastName, FirstName; LastName, FirstName" format
     * to CSL-JSON author array format
     * @param {string} authorString - Author string in VSpace format
     * @returns {Array} Array of author objects with family and given names
     */
    function parseAuthors(authorString) {
        if (!authorString || authorString.trim() === '') {
            return [];
        }

        var authors = [];
        var authorParts = authorString.split(';');

        for (var i = 0; i < authorParts.length; i++) {
            var author = authorParts[i].trim();
            if (author === '') continue;

            var nameParts = author.split(',');
            if (nameParts.length >= 2) {
                authors.push({
                    family: nameParts[0].trim(),
                    given: nameParts[1].trim()
                });
            } else {
                authors.push({
                    literal: author
                });
            }
        }

        return authors;
    }

    /**
     * Parses editors string using same format as authors
     * @param {string} editorsString - Editors string
     * @returns {Array} Array of editor objects
     */
    function parseEditors(editorsString) {
        return parseAuthors(editorsString);
    }

    /**
     * Converts Virtual Spaces Reference object to CSL-JSON format
     * @param {Object} reference - Reference object with VSpace fields
     * @returns {Object} CSL-JSON formatted object
     */
    function convertToCslJson(reference) {
        var cslItem = {
            id: reference.id || 'ref-' + Date.now(),
            type: typeMapping[reference.type] || 'document'
        };

        if (reference.title) {
            cslItem.title = reference.title;
        }

        if (reference.author) {
            cslItem.author = parseAuthors(reference.author);
        }

        if (reference.year) {
            var year = parseInt(reference.year, 10);
            if (!isNaN(year)) {
                cslItem.issued = {
                    'date-parts': [[year]]
                };
            }
        }

        if (reference.journal) {
            cslItem['container-title'] = reference.journal;
        }

        if (reference.url) {
            cslItem.URL = reference.url;
        }

        if (reference.volume) {
            cslItem.volume = reference.volume;
        }

        if (reference.issue) {
            cslItem.issue = reference.issue;
        }

        if (reference.pages) {
            cslItem.page = reference.pages;
        }

        if (reference.editors) {
            cslItem.editor = parseEditors(reference.editors);
        }

        if (reference.note) {
            cslItem.note = reference.note;
        }

        return cslItem;
    }

    /**
     * Formats a single reference using citation.js
     * @param {Object} reference - Reference object
     * @param {string} style - Citation style ('apa', 'chicago', 'mla')
     * @returns {string} Formatted HTML string
     */
    function formatReference(reference, style) {
        try {
            var Cite = require('citation-js');
            var cslData = convertToCslJson(reference);
            var cite = new Cite(cslData);

            var templateMap = {
                'apa': 'apa',
                'chicago': 'chicago-fullnote-bibliography',
                'mla': 'modern-language-association'
            };

            var template = templateMap[style] || 'apa';

            return cite.format('bibliography', {
                format: 'html',
                template: template,
                lang: 'en-US'
            });
        } catch (e) {
            console.error('Citation formatting error:', e);
            // Fallback to basic formatting
            return formatFallback(reference);
        }
    }

    /**
     * Fallback formatting when citation.js fails
     * @param {Object} reference - Reference object
     * @returns {string} Basic formatted HTML
     */
    function formatFallback(reference) {
        var parts = [];
        if (reference.author) parts.push(reference.author);
        if (reference.year) parts.push('(' + reference.year + ')');
        if (reference.title) parts.push('<em>' + escapeHtml(reference.title) + '</em>');
        if (reference.journal) parts.push(escapeHtml(reference.journal));
        if (reference.volume) parts.push(reference.volume);
        if (reference.issue) parts.push('(' + reference.issue + ')');
        if (reference.pages) parts.push(reference.pages);
        if (reference.url) parts.push('<a href="' + escapeHtml(reference.url) + '" target="_blank">' + escapeHtml(reference.url) + '</a>');

        return '<p>' + parts.join('. ') + '.</p>';
    }

    /**
     * Escapes HTML special characters
     * @param {string} text - Text to escape
     * @returns {string} Escaped text
     */
    function escapeHtml(text) {
        if (!text) return '';
        var div = document.createElement('div');
        div.textContent = text;
        return div.innerHTML;
    }

    /**
     * Formats all references in a BiblioBlock
     * @param {Array} references - Array of reference objects
     * @param {string} style - Citation style
     * @returns {string} Formatted HTML for all references
     */
    function formatBibliography(references, style) {
        var html = '';
        for (var i = 0; i < references.length; i++) {
            html += '<div class="formatted-reference">' +
                    formatReference(references[i], style) +
                    '</div>';
        }
        return html;
    }

    return {
        convertToCslJson: convertToCslJson,
        formatReference: formatReference,
        formatBibliography: formatBibliography,
        parseAuthors: parseAuthors
    };

})();
